Marlin/src/feature/fwretract.cpp

```diff
@@ -130,47 +126,39 @@ void FWRetract::retract(const bool retracting
 
   // The current position will be the destination for E and Z moves
   set_destination_from_current();
 
   if (retracting) {
     // Retract by moving from a faux E position back to the current E position
-    feedrate_mm_s = retract_feedrate_mm_s;
-    destination[E_AXIS] -= base_retract * renormalize;
+    feedrate_mm_s = retract_feedrate_mm_s * feedrate_scale;
+    current_retract[active_extruder] = base_retract * renormalize;
     prepare_move_to_destination();                        // set_current_to_destination
 
     // Is a Z hop set, and has the hop not yet been done?
-    if (retract_zlift > 0.01 && !hop_amount) {            // Apply hop only once
-      hop_amount += retract_zlift;                        // Add to the hop total (again, only once)
-      destination[Z_AXIS] += retract_zlift;               // Raise Z by the zlift (M207 Z) amount
-      feedrate_mm_s = planner.max_feedrate_mm_s[Z_AXIS];  // Maximum Z feedrate
+    if (retract_zlift > 0.01 && !current_hop) {           // Apply hop only once
+      current_hop += retract_zlift;                       // Add to the hop total (again, only once)
+      feedrate_mm_s = planner.max_feedrate_mm_s[Z_AXIS] * feedrate_scale;  // Maximum Z feedrate
       prepare_move_to_destination();                      // Raise up, set_current_to_destination
     }
   }
   else {
     // If a hop was done and Z hasn't changed, undo the Z hop
-    if (hop_amount) {
-      current_position[Z_AXIS] += hop_amount;             // Restore the actual Z position
-      SYNC_PLAN_POSITION_KINEMATIC();                     // Unspoof the position planner
-      feedrate_mm_s = planner.max_feedrate_mm_s[Z_AXIS];  // Z feedrate to max
+    if (current_hop) {
+      current_hop = 0.0;
+      feedrate_mm_s = planner.max_feedrate_mm_s[Z_AXIS] * feedrate_scale;  // Z feedrate to max
       prepare_move_to_destination();                      // Lower Z, set_current_to_destination
-      hop_amount = 0.0;                                   // Clear the hop amount
     }
 
-    destination[E_AXIS] += (base_retract + (swapping ? swap_retract_recover_length : retract_recover_length)) * renormalize;
-    feedrate_mm_s = swapping ? swap_retract_recover_feedrate_mm_s : retract_recover_feedrate_mm_s;
+    current_retract[active_extruder] = 0.0;
+    feedrate_mm_s = (swapping ? swap_retract_recover_feedrate_mm_s : retract_recover_feedrate_mm_s) * feedrate_scale;
     prepare_move_to_destination();                        // Recover E, set_current_to_destination
   }
 
   feedrate_mm_s = old_feedrate_mm_s;                      // Restore original feedrate
-  current_position[Z_AXIS] = old_z;                       // Restore Z and E positions
-  current_position[E_AXIS] = old_e;
-  SYNC_PLAN_POSITION_KINEMATIC();                         // As if the move never took place
-
   retracted[active_extruder] = retracting;                // Active extruder now retracted / recovered
```

Comment:
To make the hop and/or E move happen right away (instead of being combined with the next move) I believe there should be a move to the `current_position` inserted here.